### ISSUE TITLE
`TokenConfig` can find unexpected resource attributes

### ISSUE THREAD

**Issue by Eranziel** [foundryvtt/foundryvtt#10007]
### What happened?

I believe there is a logical error in the way `TokenConfig` gets the tracked attributes for a token. As a result, when data models are defined for the actor types in a system, `TokenConfig` will get and show an aggregate list of attributes from _all_ actor types - even attributes which do not exist on the token's actor.

For example, like this. This token's actor's data model does not include the `bond_state.stress` or `bond_state.xp` fields, those only exist on one of the four actor types in the system. Likewise, there are several single-value attributes listed which do not exist on this token's actor's data model, but do on other actor types.
![image](https://github.com/foundryvtt/foundryvtt/assets/49248157/602d8380-dde0-46d3-a040-b3934bd52e5d)

The root cause looks like it is here, in `foundry.js`:
```js
class TokenConfig {
  ...
  async getData(options={}) {
    const alternateImages = await this._getAlternateTokenImages();
    const attributeSource = this.actor?.system instanceof foundry.abstract.DataModel
      ? this.actor?.type         // If actor.system is a DataModel, *don't* pass it to getTrackedAttributes?
      : this.actor?.system;
    const attributes = TokenDocument.implementation.getTrackedAttributes(attributeSource);
```
This causes the actor's `type` string to be passed to `getTrackedAttributes`, instead of passing `actor.system`. Then, inside `getTrackedAttributes`:
- "Case 1", which examines a data model's schema for attributes, gets skipped because it doesn't have a data model to work with.
- Same with cases 2 and 3.
- Case 4 is reached, which then recursively calls `getTrackedAttributes` for every actor data model defined.

My guess is that the intent was for the ternary outputs to be the other way around:
```js
  async getData(options={}) {
    const alternateImages = await this._getAlternateTokenImages();
    const attributeSource = this.actor?.system instanceof foundry.abstract.DataModel
      ? this.actor?.system
      : this.actor?.type;
    const attributes = TokenDocument.implementation.getTrackedAttributes(attributeSource);
```
Simulating this by setting a breakpoint and assigning `attributeSource = this.actor.system` before the call to `getTrackedAttributes` gives the expected resources:
![image](https://github.com/foundryvtt/foundryvtt/assets/49248157/c42937dc-2ed0-46cf-886d-01c2d9dba715)

### What ways of accessing Foundry can you encounter this issue in?

- [X] Native App (Electron)
- [X] Chrome
- [X] Firefox
- [ ] Safari
- [ ] Other

### Reproduction Steps

This is a bit tricky! It is easily reproducible on my dev branch of Lancer by opening token config on any token. However, there isn't yet a public release with data models configured. I am happy to assist by either helping someone on the team to clone and build the branch, or I can send someone a zip of the built system files. You can contact me on Discord `@eranziel`.

Otherwise, it should be reproducible by:
- Defining 2 actor types with some differing fields in their schemas.
- Creating an actor of either type.
- Create a token from the actor.
- Open token configuration, go to resources tab, see all attribute fields from both schemas.

### What core version are you reporting this for?

Version 11 Stable (308)

### Bug Checklist

- [X] The issue occurs while all Modules are disabled

**Comment by aaclayton** [comment]
@Fyorl perhaps you recall some of the history here? We used to always pass a string `type` into `TokenDocument.implementation.getTrackedAttributes(this.actor?.type)`, but then we made this change to sometimes pass the whole `system` objects (for non-DataModel systems). 

I expect our intention is that we want DataModel enabled systems to get all their tracked attributes via `TokenDocument._getTrackedAttributesFromSchema`, but `TokenDocument.getTrackedAttributes` won't end up going down this path if passed a string.

**Comment by Fyorl** [comment]
> I expect our intention is that we want DataModel enabled systems to get all their tracked attributes via `TokenDocument._getTrackedAttributesFromSchema`, but `TokenDocument.getTrackedAttributes` won't end up going down this path if passed a string.

I believe the intention was that DataModel-enabled systems should use the new `CONFIG.Actor.trackableAttributes` since there is no other way to include derived attributes, see [here](https://github.com/foundryvtt/foundryvtt/issues/8726#issuecomment-1505777803).

**Comment by aaclayton** [comment]
I see, thanks. So I suppose the OP is in a position where they have a `DataModel` in the system, but no `CONFIG.Actor.trackableAttributes` defined. In such a circumstance we end up falling through the cracks and not using the actual data model instance that exists on the actor. 

I suspect there's some improvement we can make here, but probably not in the scope of V11. Re-assigning this to V12.

@Eranziel in the mean-time you can work through this by defining the `CONFIG.Actor.trackableAttributes` data structure for your actor types which will explicitly declare which attributes should be trackable choices in this context.

**Comment by Fyorl** [comment]
> I suspect there's some improvement we can make here, but probably not in the scope of V11. Re-assigning this to V12.

Yes you are right, we could probably improve the logic used in `TokenConfig` to determine what information it sends to `TokenDocument.getTrackedAttributes`.

**Comment by Eranziel** [comment]
> I see, thanks. So I suppose the OP is in a position where they have a `DataModel` in the system, but no `CONFIG.Actor.trackableAttributes` defined. In such a circumstance we end up falling through the cracks and not using the actual data model instance that exists on the actor.
> 
> I suspect there's some improvement we can make here, but probably not in the scope of V11. Re-assigning this to V12.
> 
> @Eranziel in the mean-time you can work through this by defining the `CONFIG.Actor.trackableAttributes` data structure for your actor types which will explicitly declare which attributes should be trackable choices in this context.

This might start to tread outside the scope of the issue title, so apologies in advance. ;) I had tried using `CONFIG.Actor.trackableAttributes` before, exactly because I was using some derived fields that I wanted to be trackable. However, it wasn't having the result I needed.

Some background context. For Lancer, there are several attributes where we only wanted to save the value and not the min/max. For example, max HP is fully derivable from the items and other stats on an actor. The data models were implemented so that the source data for HP was a simple `number`, and `prepareData` etc... turned that into a `{min, max, value}` object. `max` was calculated using other stats and items, and `value` took the value of `hp` from source.

This all worked fine, _except_ for token bar functionality. They would always show up as single values in `TokenConfig`, and the bar value was not editable from the token HUD. I suspect this is because `actor.system.hp` _did_ exist in the source model, but it was a `number` instead of an object with `{min, max, value}`. In other words, `getTrackableAttributes` was using the source model instead of the derived system model.

I tried several different solutions, one of which was explicitly setting them in `CONFIG.Actor.trackableAttributes`. IIRC that resulted in them showing up as attributes in `TokenConfig`, but they were still not editable for the token HUD. I can confirm tonight once I'm back at my Lancer dev env, if you like.

The solution I am currently working with is to ditch the fancy footwork and just define those fields as `{min, max, value}` objects all the way through, from `template.json` up. (Incidentally, having to maintain a `template.json` when you have data models fully defined for all actors and items is tedious and redundant. It would be great if the schema was collected from the data models if `template.json` isn't present.) This results in a bit of superfluous data being transmitted back and forth and being saved to the DB, but those tiny savings were causing more headaches than they're worth due to the various special workarounds required, like this issue with token bars.

**Comment by Eranziel** [comment]
The solution discussed in that last paragraph is working, to be clear. However, it has the oddity of showing the trackable attributes for _all_ actor types, not just the one related to the token. That's what prompted me to make this bug report.

**Comment by Fyorl** [comment]
> This all worked fine, _except_ for token bar functionality. They would always show up as single values in `TokenConfig`, and the bar value was not editable from the token HUD. I suspect this is because `actor.system.hp` _did_ exist in the source model, but it was a `number` instead of an object with `{min, max, value}`. In other words, `getTrackableAttributes` was using the source model instead of the derived system model.

> I tried several different solutions, one of which was explicitly setting them in `CONFIG.Actor.trackableAttributes`. IIRC that resulted in them showing up as attributes in `TokenConfig`, but they were still not editable for the token HUD. I can confirm tonight once I'm back at my Lancer dev env, if you like.

That would be helpful, yes. I would like to go over the different `trackableAttributes` configurations you tried and see why they didn't work, and if we can make them work (assuming it's not just a bug, which is possible.)
